### PR TITLE
3/n Move accelerator into Strategy

### LINKED ISSUE #10648: [Accelerator refactor] Move Accelerator into Strategy
## Proposed refactor

Part 3 of Accelerator and Plugin refactor #10416 

### Motivation

Moving towards stable version

After step 2 Move Precision Plugin into TTP Precision Plugins should be part of Training Type Plugins #7324
Accelerator is not the routing layer for strategy and precision anymore, optimizer related logic, steps, hooks all moved in to strategy. 

Accelerator only have device information - strategy should own accelerator. Reduce the code complexity and improve code maintainability  

### Pitch
move accelerator into ttp/strategy as device_plugin (similar to checkpoint_io) and updating logic in accelerator-connector, training, loops accordingly

[RFC] Should we have a new name for Accelerator?

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @akihironitta @kaushikb11 @ananthsub

### PR DESCRIPTION
**Pull request by four4fish**
## What does this PR do?
Last step of Move accelerator into Strategy

Fixes #10648 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 1]
```diff
@@ -405,11 +407,30 @@ def training_type_plugin(self) -> TrainingTypePlugin:
...
-
+        if (
+            # handle custom strategy with custom accelerator
+            (
+                isinstance(self.strategy, TrainingTypePlugin)
+                and (self.strategy.accelerator is None or not isinstance(self.strategy.accelerator, Accelerator))
+            )
+            or not isinstance(self.strategy, TrainingTypePlugin)
+        ):
+            self._training_type_plugin.accelerator = self.select_accelerator()
```
Comment (four4fish) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:433]: @awaelchli this is really ugly, but this is the minimal impact way I can think about to handle custom plugins. How about we refactor/rewrite [Accelerator_connector](https://github.com/PyTorchLightning/pytorch-lightning/issues/10422) after this and keep logic as same as possible compare to before for now?

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 2]
```diff
@@ -405,11 +407,17 @@ def training_type_plugin(self) -> TrainingTypePlugin:
...
+        ):
+            precision_plugin = self.precision_plugin
+            if precision_plugin is not None:
+                self._training_type_plugin._precision_plugin = precision_plugin
         self._training_type_plugin_resolved = True
-
+        if (hasattr(self.strategy, "accelerator") and self.strategy.accelerator is None) or not hasattr(
+            self.strategy, "accelerator"
+        ):
+            self._training_type_plugin.accelerator = self.select_accelerator()
```
Comment (four4fish) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:420]: @awaelchli this is really ugly, but this is the minimal impact way I can think about to handle custom plugins. How about we refactor/rewrite Accelerator_connector after this and keep logic as same as possible compare to before for now?

File: tests/accelerators/test_accelerator_connector.py [hunk 3]
```diff
@@ -1029,10 +1034,13 @@ def test_unsupported_tpu_choice(monkeypatch):
     with pytest.raises(MisconfigurationException, match=r"accelerator='tpu', precision=64\)` is not implemented"):
         Trainer(accelerator="tpu", precision=64)
 
-    with pytest.warns(UserWarning, match=r"accelerator='tpu', precision=16\)` but native AMP is not supported"):
-        Trainer(accelerator="tpu", precision=16)
-    with pytest.warns(UserWarning, match=r"accelerator='tpu', precision=16\)` but apex AMP is not supported"):
-        Trainer(accelerator="tpu", precision=16, amp_backend="apex")
+    with pytest.raises(ValueError, match="TPUAccelerator` can only be used with a `SingleTPUPlugin`"):
```
Comment (four4fish) [inline tests/accelerators/test_accelerator_connector.py:1037]: reviewers: the TPU misconfig check has been moved from TPUAccelerator setup to accelerator_connector, so it will get checked when trainer initialize

File: pytorch_lightning/accelerators/tpu.py [hunk 4]
```diff
@@ -29,25 +25,6 @@
...
-                f"The `TPUAccelerator` can only be used with a `TPUPrecisionPlugin`,"
-                f" found: {self.training_type_plugin.precision_plugin}."
-            )
-        if not isinstance(self.training_type_plugin, (SingleTPUPlugin, TPUSpawnPlugin)):
-            raise ValueError(
-                "The `TPUAccelerator` can only be used with a `SingleTPUPlugin` or `TPUSpawnPlugin,"
-                f" found {self.training_type_plugin}."
-            )
-        return super().setup(trainer)
-
```
Comment (four4fish) [inline pytorch_lightning/accelerators/tpu.py:50]: reviewers: this has been moved to the end of the accelerator_connector, will be called at the end of the accelerator_connector init

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 5]
```diff
@@ -394,22 +395,23 @@ def precision_plugin(self) -> PrecisionPlugin:
...
+        if (
+            isinstance(self.strategy, TrainingTypePlugin) and self.strategy._precision_plugin is None
+        ) or not isinstance(self.strategy, TrainingTypePlugin):
+            precision_plugin = self.precision_plugin
+            if precision_plugin is not None:
+                self._training_type_plugin.precision_plugin = precision_plugin
+        if (isinstance(self.strategy, TrainingTypePlugin) and self.strategy.accelerator is None) or not isinstance(
+            self.strategy, TrainingTypePlugin
+        ):
+            self._training_type_plugin.accelerator = self.select_accelerator()
```
Comment (four4fish) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:414]: @awaelchli this is really ugly, but this is the minimal impact way I can think about to handle custom plugins. How about we refactor/rewrite Accelerator_connector after this and keep logic as same as possible compare to before for now?

File: pytorch_lightning/accelerators/tpu.py (hunk 4 above)
Comment (ananthsub) [reply pytorch_lightning/accelerators/tpu.py:50]: why not move the gpu/cpu acelerator checks to the accelerator connector as well?

**Review by ananthsub — COMMENTED** [review]
nit: updating references for `trainer.accelerator.training_type_plugin` to `trainer.training_type_plugin` ((e.g. in files test_deepspeed_plugin and test_ddp_plugin_wih_comm_hook) could be a separate PR to make this one smaller and easier to review

File: pytorch_lightning/accelerators/accelerator.py [hunk 6]
```diff
@@ -12,14 +12,11 @@
...
 
 import torch
-from torch.nn import Module
 
 import pytorch_lightning as pl
-from pytorch_lightning.plugins.precision import PrecisionPlugin
-from pytorch_lightning.plugins.training_type import TrainingTypePlugin
 
 
 class Accelerator:
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/accelerator.py:25]: I understand this PR is to facilitate the refactor. What should the ideal Accelerator API look like? It looks like the accelerator is now not much more than the torch.device being used, vendor specific setup, and a way to get device stats (which is not a bad thing)

File: pl_examples/bug_report/bug_report_model.py [hunk 7]
```diff
@@ -57,6 +57,9 @@ def run():
         num_sanity_val_steps=0,
         max_epochs=1,
         enable_model_summary=False,
+        accelerator="cpu",
+        strategy="ddp",
+        devices=2,
```
Comment (carmocca) [inline pl_examples/bug_report/bug_report_model.py:62]: This shouldn't be changed

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 8]
```diff
@@ -797,9 +799,9 @@ def select_accelerator(self) -> Accelerator:
         else:
             acc_cls = CPUAccelerator
 
-        accelerator = acc_cls(precision_plugin=None, training_type_plugin=self.training_type_plugin)
+        accelerator = acc_cls()
         # transfer ownership of the plugins to the accelerator
-        self._training_type_plugin = proxy(self.training_type_plugin)
+        # self._training_type_plugin = proxy(self.training_type_plugin)
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:804]: Remove this?

File: pytorch_lightning/trainer/trainer.py [hunk 9]
```diff
@@ -1671,11 +1672,11 @@ def _on_exception(self) -> None:
 
     @property
     def accelerator(self) -> Accelerator:
-        return self._accelerator_connector.accelerator
+        return self.training_type_plugin.accelerator
 
     @property
     def training_type_plugin(self) -> TrainingTypePlugin:
-        return self.accelerator.training_type_plugin
+        return self._accelerator_connector.training_type_plugin
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1679]:
```suggestion
        return self.training_type_plugin
```

File: pytorch_lightning/accelerators/accelerator.py (hunk 6 above)
Comment (four4fish) [reply pytorch_lightning/accelerators/accelerator.py:25]: @ananthsub @justusschock, I agree that current Accelerator is more like a Device class and it's a good thing to have :) And more device related check/logic could be moved into this Accelerator subclasses after this 3/n. Especially as @justusschock mentioned logic like has_tpu(). Or mis configs/error specific to one Accelerator like cpu `num_processes` 
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/accelerator_connector.py#L262-L264




>Make this the only source of truth for the used device type (e.g either replace the enum, or make it a property of this class) as I find it kind of confusing, that we check different things here and they could potentially diverge (although they shouldn't :)

Totally agree, we are planning on deprecating the trainer._device_type from and unify the check for device type to check accelerator class [Remove trainer._device_type in favor of check Accelerator class](https://github.com/PyTorchLightning/pytorch-lightning/issues/11002) and [Generalize internal checks for precision plugin type, training type, accelerator type](https://github.com/PyTorchLightning/pytorch-lightning/issues/10821) 
+1 make device_type a property of Accelerator class and deprecate Accelerator enum. (I think there will be circle import issue to solve before we can deprecate accelerator_enum)

File: pytorch_lightning/accelerators/tpu.py (hunk 4 above)
Comment (four4fish) [reply pytorch_lightning/accelerators/tpu.py:50]: It's a bit different. This TPU check is checking what strategies and precision plugins to use when device is TPU, which should happens before in plugin selection phase like accelerator_connector.

For device specific checks like what's in cpu/gpu now, I think it will make sense to slim the accelerator_connector and moving to accelerator subclass.

File: pytorch_lightning/accelerators/gpu.py [hunk 10]
```diff
@@ -30,22 +30,19 @@
...
-            raise MisconfigurationException(
-                f"Device should be GPU, got {self.training_type_plugin.root_device} instead"
-            )
-        torch.cuda.set_device(self.training_type_plugin.root_device)
+        if "cuda" not in str(root_device):
+            raise MisconfigurationException(f"Device should be GPU, got {self.root_device} instead")
+        torch.cuda.set_device(root_device)
 
     def setup(self, trainer: "pl.Trainer") -> None:
+        # TODO refactor input from trainer to local_rank @four4fish
```
[the remark this answers is not in the document]
Comment (four4fish) [reply pytorch_lightning/accelerators/gpu.py:44]: I don't think so. root_device not necessarily equal to root_device. Some strategies has root_device to be parallel_device[local_rank] like ddp, but some plugins has different implementation like dp and ddp2 (root_device always be device[0]), and single device doesn't have local rank property.

File: pytorch_lightning/trainer/trainer.py [hunk 11]
```diff
@@ -451,6 +451,7 @@ def __init__(
             amp_level,
             plugins,
         )
+        self._accelerator_connector.training_type_plugin
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:454]: note to self: need to find a way to avoid this. at the very least, this should be done within the connector

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 12]
```diff
@@ -1016,3 +1014,18 @@ def _is_slurm_managing_tasks(self) -> bool:
         total_requested_devices = (self.num_gpus or self.num_processes) * self.num_nodes
         num_slurm_tasks = int(os.environ["SLURM_NTASKS"], 0)
         return num_slurm_tasks == total_requested_devices
+
+    def _check_tpu_mis_config(self) -> None:
+        # TODO moved from TPUAccelerator when refactor accelerator. Revisit when refactor
+        # accelerator_connector @four4fish
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:1025]: @four4fish I would avoid such TODOs (from experience with other refactors) and rather note them to the related issue.

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:414]: Thank you. Yes whatever makes the tests pass. At this point I have no clue anymore.

File: tests/accelerators/test_cpu.py [hunk 13]
```diff
@@ -16,9 +16,10 @@
 
 def test_restore_checkpoint_after_pre_dispatch_default():
     """Assert default for restore_checkpoint_after_pre_dispatch is False."""
-    plugin = SingleDevicePlugin(torch.device("cpu"))
-    accelerator = CPUAccelerator(training_type_plugin=plugin, precision_plugin=PrecisionPlugin())
-    assert not accelerator.training_type_plugin.restore_checkpoint_after_pre_dispatch
+    plugin = SingleDevicePlugin(
+        accelerator=CPUAccelerator(), device=torch.device("cpu"), precision_plugin=PrecisionPlugin()
+    )
+    assert not plugin.restore_checkpoint_after_pre_dispatch
```
Comment (carmocca) [inline tests/accelerators/test_cpu.py:22]: Dupe

```suggestion
```

File: tests/accelerators/test_accelerator_connector.py [hunk 14]
```diff
@@ -408,12 +415,11 @@ class Accel(Accelerator):
     class Prec(PrecisionPlugin):
         pass
 
-    class TrainTypePlugin(SingleDevicePlugin):
```
Comment (carmocca) [inline tests/accelerators/test_accelerator_connector.py:411]: Why this change?

File: tests/accelerators/test_accelerator_connector.py [hunk 15]
```diff
@@ -397,7 +397,14 @@ def creates_processes_externally(self) -> bool:
...
 @mock.patch.dict(
     os.environ,
-    {"SLURM_NTASKS": "2", "SLURM_JOB_NAME": "SOME_NAME", "SLURM_NODEID": "0", "LOCAL_RANK": "0", "SLURM_LOCALID": "0"},
+    {
+        "SLURM_NTASKS": "2",
+        "SLURM_JOB_NAME": "SOME_NAME",
+        "SLURM_NODEID": "0",
+        "LOCAL_RANK": "0",
+        "SLURM_PROCID": "0",
+        "SLURM_LOCALID": "0",
```
Comment (carmocca) [inline tests/accelerators/test_accelerator_connector.py:406]: Removing this comma will undo this unrelated change

```suggestion
        "SLURM_LOCALID": "0"
```

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 16]
```diff
@@ -1016,3 +1014,23 @@ def _is_slurm_managing_tasks(self) -> bool:
...
+        if isinstance(self.accelerator, TPUAccelerator):
+            if not isinstance(self.training_type_plugin.precision_plugin, TPUPrecisionPlugin):
+                raise ValueError(
+                    f"The `TPUAccelerator` can only be used with a `TPUPrecisionPlugin`,"
+                    f" found: {self.training_type_plugin.precision_plugin}."
+                )
+            if not isinstance(self.training_type_plugin, (SingleTPUPlugin, TPUSpawnPlugin)):
+                raise ValueError(
+                    "The `TPUAccelerator` can only be used with a `SingleTPUPlugin` or `TPUSpawnPlugin,"
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:1034]:
```suggestion
                    "The `TPUAccelerator` can only be used with a `SingleTPUPlugin` or `TPUSpawnPlugin`,"
```

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 17]
```diff
@@ -997,14 +996,13 @@ def _set_distrib_type_if_training_type_plugin_passed(self):
...
+        conditions match:
 
-            - A SLURM cluster is detected
-            - A distributed plugin is being used
-            - The process is not launching in interactive mode
-            - The number of tasks in SLURM matches the requested number of devices and nodes in the Trainer
+        - A SLURM cluster is detected
+        - A distributed plugin is being used
+        - The process is not launching in interactive mode
+        - The number of tasks in SLURM matches the requested number of devices and nodes in the Trainer
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:1005]: Keep the space before the bullet points?

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 18]
```diff
@@ -57,13 +61,25 @@ def __init__(
...
         return self._checkpoint_io
 
     @property
     def precision_plugin(self) -> PrecisionPlugin:
-        return self._precision_plugin
+        return self._precision_plugin if self._precision_plugin is not None else PrecisionPlugin()
+
+    @precision_plugin.setter
+    def precision_plugin(self, precision_plugin: Optional[PrecisionPlugin]) -> None:
+        self._precision_plugin = precision_plugin
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:82]: Adding this in #11071

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 17 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:1005]: this was modified by the weirdo pre-commit bot

File: docs/source/extensions/plugins.rst [hunk 19]
```diff
@@ -80,10 +80,7 @@ can then be passed into the Trainer directly or via a (custom) accelerator:
     trainer = Trainer(strategy=CustomDDPPlugin(), plugins=[CustomPrecisionPlugin()])
 
     # fully custom accelerator and plugins
```
Comment (ananthsub) [inline docs/source/extensions/plugins.rst:82]: Custom plugins are dropped from the example?

File: pytorch_lightning/__init__.py [hunk 20]
```diff
@@ -13,12 +13,13 @@
...
     _logger.propagate = False
 
+from pytorch_lightning.accelerators import Accelerator  # noqa: E402
 from pytorch_lightning.callbacks import Callback  # noqa: E402
 from pytorch_lightning.core import LightningDataModule, LightningModule  # noqa: E402
 from pytorch_lightning.trainer import Trainer  # noqa: E402
 from pytorch_lightning.utilities.seed import seed_everything  # noqa: E402
 
-__all__ = ["Trainer", "LightningDataModule", "LightningModule", "Callback", "seed_everything"]
+__all__ = ["Accelerator", "Trainer", "LightningDataModule", "LightningModule", "Callback", "seed_everything"]
```
Comment (ananthsub) [inline pytorch_lightning/__init__.py:22]: Why this change?

File: pytorch_lightning/accelerators/accelerator.py [hunk 21]
```diff
@@ -31,76 +28,21 @@ class Accelerator:
...
-        To get the potentially wrapped model use :attr:`Accelerator.model`
-        """
-        return self.training_type_plugin.lightning_module
-
-    @property
-    def root_device(self) -> torch.device:
-        """Returns the root device."""
-        return self.training_type_plugin.root_device
-
-    def teardown(self) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/accelerator.py:98]: Since the interface supports a setup, we should have a teardown

File: pytorch_lightning/__init__.py (hunk 20 above)
Comment (awaelchli) [reply pytorch_lightning/__init__.py:22]: to annotate the types in the plugins with "pl.Accelerator". Otherwise there appear to be circular imports unfortunately. Haven't found a different solution yet.

Comment (justusschock) [reply pytorch_lightning/__init__.py:22]: so annotating with `"pl.accelerators.accelerator.Accelerator"` doesn't work? This might be a bit longer, but since we don't import it directly, that should be fine I think

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 17 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:1005]: The weirdo https://github.com/myint/docformatter project. pre-commit just obeys what we tell it to do.

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:1005]: Note that the bot did this in reaction to a change by a human. Probably moving "Returns ``True..." to the first paragraph was the cause.

File: pytorch_lightning/trainer/trainer.py [hunk 22]
```diff
@@ -1124,7 +1125,7 @@ def _run(
         {Trainer.fit} or {Trainer.test} or {Trainer.predict}  ||
                                 |                             ||
                          spawn processes                      ||
-               {self.accelerator.setup_environment}           ||
+               {self.training_type_plugin.setup_environment}           ||
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1128]: This needs adjustment

File: pytorch_lightning/accelerators/accelerator.py (hunk 21 above)
Comment (carmocca) [reply pytorch_lightning/accelerators/accelerator.py:98]: May I suggest we only add hooks when they have actual implementations? Adding just because is how we've ended up with that many deprecated hooks.

File: pytorch_lightning/__init__.py (hunk 20 above)
Comment (carmocca) [reply pytorch_lightning/__init__.py:22]: `__all__` is just to indicate what's imported with `from pytorch_lightning import *`

It shouldn't relate to circular imports at all

File: pytorch_lightning/accelerators/accelerator.py (hunk 21 above)
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator.py:98]: After this refactor, we will have the opportunity to split out device specific teardown that currently lives in the ddp/spawn plugin teardown and move it to the accelerator teardown. So either way, it will be sorted out :)

Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator.py:98]: https://github.com/PyTorchLightning/pytorch-lightning/blob/ffb1a754af5301608c71955c12468216e920af51/pytorch_lightning/plugins/training_type/ddp.py#L504-L508

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 18 above)
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:82]: Thank you @carmocca, will rebase if 11071 merge first. The accelerator_connector logic change in this PR need precision setters for now

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 12 above)
Comment (four4fish) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:1025]: Yeah make sense, will remove

File: pytorch_lightning/trainer/trainer.py (hunk 11 above)
Comment (four4fish) [reply pytorch_lightning/trainer/trainer.py:454]: Yeah agree, at least we should wet it in accelerator_connector

File: pytorch_lightning/__init__.py (hunk 20 above)
Comment (awaelchli) [reply pytorch_lightning/__init__.py]: okay whatever works

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 23]
```diff
@@ -42,12 +42,16 @@ class TrainingTypePlugin(ABC):
     loop."""
 
     def __init__(
-        self, checkpoint_io: Optional[CheckpointIO] = None, precision_plugin: Optional[PrecisionPlugin] = None
+        self,
+        accelerator: Optional["pl.accelerators.accelerator.Accelerators.accelerator.Accelerator"] = None,
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:46]: I legitimally laughed at this 😂😂😂

Did find+replace trick you?

Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:46]: how is this possible ...

Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:46]:
```suggestion
        accelerator: Optional["pl.accelerators.accelerator.Accelerator"] = None,
```

File: docs/source/extensions/accelerators.rst [hunk 24]
```diff
@@ -25,11 +25,10 @@ One to handle differences from the training routine and one to handle different
     from pytorch_lightning.accelerators import GPUAccelerator
     from pytorch_lightning.plugins import NativeMixedPrecisionPlugin, DDPPlugin
 
-    accelerator = GPUAccelerator(
-        precision_plugin=NativeMixedPrecisionPlugin(precision=16, device="cuda"),
-        training_type_plugin=DDPPlugin(),
-    )
-    trainer = Trainer(accelerator=accelerator)
+    accelerator = GPUAccelerator()
+    precision_plugin = NativeMixedPrecisionPlugin(precision=16, device="cuda")
```
[the remark this answers is not in the document]
Comment (four4fish) [reply docs/source/extensions/accelerators.rst:29]: Yeah make sense. Will create a issue to make the device param optional in native_mixed precision plugins and clean up here

File: docs/source/extensions/plugins.rst (hunk 19 above)
Comment (four4fish) [reply docs/source/extensions/plugins.rst:82]: Will the example above sufficient?

File: tests/accelerators/test_accelerator_connector.py (hunk 14 above)
Comment (four4fish) [reply tests/accelerators/test_accelerator_connector.py:411]: Was just for my own testing purpose. Good spotting :) will change it back

File: docs/source/extensions/plugins.rst (hunk 19 above)
Comment (four4fish) [reply docs/source/extensions/plugins.rst:82]: added

**Comment by four4fish** [comment]
Reviewers: most issue has been addressed, thank you for feedbacks and discussions!
Follow up for this PR:
1. Teardown hook, issue filed : https://github.com/PyTorchLightning/pytorch-lightning/issues/11091
2. Simplify NativeMixedPrecisionPlugin init param and remove device as required param. It's probably less trivial then it looks as it will require accelerator_connector changes.

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 25]
```diff
@@ -425,6 +443,7 @@ def teardown(self) -> None:
 
         It is the right place to release memory and free other resources.
         """
+        self._move_optimizer_state(torch.device("cpu"))
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:446]: Note, if I read correctly this was previously only done for the GPU accelerator. Should be fine, just pointing it out in case.

File: pytorch_lightning/trainer/trainer.py (hunk 11 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:454]: actually, since it is no longer a property I think this line is now redundant. It does not do anything and let's try to remove it.

**Review by awaelchli — APPROVED** [review]
Very nice!

File: pytorch_lightning/trainer/trainer.py (hunk 11 above)
Comment (four4fish) [reply pytorch_lightning/trainer/trainer.py]: Oh! good catch, it was added for debugging. Now accelerator_connector will initialized the ttp

File: pytorch_lightning/trainer/trainer.py [hunk 26]
```diff
@@ -451,7 +451,7 @@ def __init__(
             amp_level,
             plugins,
         )
-        self.logger_connector = LoggerConnector(self, log_gpu_memory)
+        self.logger_connector: LoggerConnector = LoggerConnector(self, log_gpu_memory)
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:454]: nit: not required

```suggestion
        self.logger_connector = LoggerConnector(self, log_gpu_memory)
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 26 above)
Comment (four4fish) [reply pytorch_lightning/trainer/trainer.py:454]: @awaelchli, did you add this to fix mypy?

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 25 above)
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:446]: I don't remember you or me did this. But probably we should keep this in GPUAccelerator teardown. How about we address it together in https://github.com/PyTorchLightning/pytorch-lightning/issues/11091?

Comment (four4fish) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:446]: I don't remember you or me did this. But probably we should keep this in GPUAccelerator teardown. How about we address it together in #11091?

File: pytorch_lightning/trainer/trainer.py (hunk 26 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:454]: unfortunately, typing breaks if this is removed.

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 25 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:446]: should be good either way. for tpu it's just redundant because the teardown runs in the spawned processes which, when they end, get completely destroyed so nothing is left in memory. 
But in the future, we may decide to bring the optimizer states back to the main process through the queue and restore them like we  do for the model weights. Anyway, this is open for discussion

File: pytorch_lightning/trainer/trainer.py (hunk 26 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:454]: I just pulled the PR, removed the annotation, and ran mypy (version 0.910). Everything is green

```suggestion
        self.logger_connector = LoggerConnector(self, log_gpu_memory)
```

Comment (four4fish) [reply pytorch_lightning/trainer/trainer.py:454]: Awesome! I saw you have a mypy PR out, should I rebase on top of that after it merges and remove this?

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:454]: You can remove this already. This PR just needs to rebase on top of #11095
